### DISCUSSION TITLE
Getting large number of thumbnails in background thread

### DISCUSSION THREAD

**Discussion by almagest21** [kingslay/KSPlayer#275] [Q&A] [answered]
Thank you for active update.

I need thumbnail images as icons for the video files to be displayed in the playlist view of my app.
I found the thumbnailImageAtCurrentTime() to get it, so after setting the file in PlayerView or PlayerLayer, I can use changeLoadState() or something to make sure the file is opened and then get the thumbnail. However, it seems to be heavy and time consuming to do one by one for a large number of files if using PlayerView. Also it is ideally done in a background thread to avoid blocking GUI.
Could you let me know if there is a better way to do this?

**Comment by kingslay** [comment, marked as answer]
> Thank you for active update.
> 
> I need thumbnail images as icons for the video files to be displayed in the playlist view of my app. I found the thumbnailImageAtCurrentTime() to get it, so after setting the file in PlayerView or PlayerLayer, I can use changeLoadState() or something to make sure the file is opened and then get the thumbnail. However, it seems to be heavy and time consuming to do one by one for a large number of files if using PlayerView. Also it is ideally done in a background thread to avoid blocking GUI. Could you let me know if there is a better way to do this?

i fixed at commit https://github.com/kingslay/KSPlayer/commit/24615a4df7723e2846f1168f576a33814d9d05a1. Can you try it and see if the problem is solved?

**Reply by almagest21** [reply]
Thank you very much. I tried to create a instance of KSPlayerLayer (which is different from Video Playback) dedicated to take thumbnails. I have took few thumbnails in same timing, it seems to be working. (it can be several hundred or more, so I need to put a limit on the number of concurrent tasks)

In addition, if possible, Could you consider to make KSMEPlayer.videoOutput.pixelBuffer as public?
I understand that there are several cases that UIImage is used, but for the most speed, it is the best way that using CVPixelBuffer as is. UIImage and converting is still slow.

And furthermore, it would be nice to be able to use AVPlayerItem in the KSAVPlayer case (perhaps it is a timing issue, but my attempt to get the display content in KSAVPlayer is nil when I get KSAVPlayer.player.currentItem not working).

**Reply by kingslay** [reply]
commit b9e14f756fdc385bf82880ca24e3fc95114f6485

**Reply by almagest21** [reply]
The public CVPixelBuffer is working perfectly! Many thanks.
